backend/module/eCampCore/test/Entity/LoginTest.php

```diff
@@ -0,0 +1,73 @@
+<?php
+
+namespace eCamp\CoreTest\Entity;
+
+use eCamp\Core\Entity\Login;
+use eCamp\Core\Entity\User;
+use eCamp\LibTest\PHPUnit\AbstractTestCase;
+use ReflectionClass;
+
+/**
+ * @internal
+ */
+class LoginTest extends AbstractTestCase {
+    public function testCreateLogin() {
+        $user = new User();
+        $login = new Login($user, 'test-password');
+
+        $this->assertTrue($login->checkPassword('test-password'));
+        $this->assertFalse($login->checkPassword('wrong-password'));
+    }
+
+    public function testPwResetKey() {
+        $user = new User();
+        $login = new Login($user, 'test-password');
+
+        $key = $login->createPwResetKey();
+        $this->assertTrue($login->checkPwResetKey($key));
+
+        $login->clearPwResetKey();
+        $this->assertFalse($login->checkPwResetKey($key));
+
+        $key = $login->createPwResetKey();
+        $login->resetPassword($key, 'new-password');
+        $this->assertTrue($login->checkPassword('new-password'));
+        $this->assertFalse($login->checkPassword('test-password'));
+
+        $key = $login->createPwResetKey();
+        $this->assertFalse($login->checkPwResetKey('wrong-key'));
+
+        $this->expectException(\Exception::class);
+        $login->resetPassword('wrong-key', 'newer-password');
+    }
+
+    public function testChangePassword() {
+        $user = new User();
+        $login = new Login($user, 'test-password');
+        $this->assertTrue($login->checkPassword('test-password'));
+
+        $login->changePassword('test-password', 'new-password');
+        $this->assertTrue($login->checkPassword('new-password'));
+        $this->assertFalse($login->checkPassword('test-password'));
+
+        $this->expectException(\Exception::class);
+        $login->changePassword('test-password', 'new-password');
+    }
+
+    public function testUpdateHashVersion() {
+        $password = (new ReflectionClass(Login::class))->getProperty('password');
+        $password->setAccessible(true);
+
+        $user = new User();
+        $login = new Login($user, 'test-password');
+
+        $key = $login->createPwResetKey();
+        $login->resetPassword($key, 'new-password', 0);
+        $this->assertTrue($login->checkPassword('new-password'));
+        $this->assertEquals('new-password', $password->getValue($login));
+
+        $login->checkPassword('new-password', true);
+        $this->assertTrue($login->checkPassword('new-password'));
+        $this->assertNotEquals('new-password', $password->getValue($login));
```

Comment:
Could you add some comments here? It was much harder to grasp what's going on compared to the above tests. Maybe for the first block something like
`// Reset password with $hashVersion = 0, i.e. cleartext`
and for the second block
`// Check and rehash password with CURRENT_HASH_VERSION, i.e. salted and hashed`